api/APIpermissions.py

```diff
@@ -1,8 +1,32 @@
+from datetime import datetime, timedelta, timezone
+
+from django.conf import settings
 from rest_framework.permissions import BasePermission
 
 from .serializers import Project
 
 
+class IsTokenAuthenticated(BasePermission):
+    code = "401 Unauthorized"
+    message = "The authentication token is not valid. It may have expired."
+
+    def has_permission(self, request, view):
+        """
+        Denies requests for users who lack a valid authentication token.
+        Similar to DRF:s own IsAuthenticated permission class.
+        Returns a boolean value that DRF will convert to an exception.
+        """
+
+        # If the existing token is older than AUTH_TOKEN_EXPIRATION, then recreate the object
+        # Give a small 10 second grace duration to give the CustomAuthToken class a change to renew the token
+        token_expiry = request.auth.created + timedelta(seconds=settings.AUTH_TOKEN_EXPIRATION + 10)
```

Comment:
This makes my head spin a little bit. Is it obvious that the `request.auth.created` is a datetime object and that the `+` operator just adds seconds to expiry? Is it a bad a idea to explicitly format `request.auth.created` to a datetime object, so that we are always sure that this will work? 

I can't find any documentation on this.

Reply:
I think that with timedelta() converting a sum of integer seconds it is quite readable, but I respect there may be a difference of opinions. I will add a comment.  For what it is worth here is Mistral Le Chat's explanation of the code.

_This line of code is adding a certain number of seconds to a timestamp to get a new timestamp in the future... The created attribute of this object is a datetime object that represents when the authentication was created, such as when the user logged in or when an authentication token was generated. ... So, in summary, this line of code is computing the expiration time of an authentication token, with an additional 10 seconds of buffer time._

Reply:
Yes, if you know that it is a timestamp, then i makes sense, but i would assume that `request.auth.created` is a string, since it usually is a json dict, but here it seems to be an object where the `created` attribute is a timestamp. 
This is not important, i just got stuck on it.